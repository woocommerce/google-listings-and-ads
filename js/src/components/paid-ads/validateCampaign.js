/**
 * External dependencies
 */
import { __, sprintf } from '@wordpress/i18n';

/**
 * @typedef {import('.~/components/types.js').CampaignFormValues} CampaignFormValues
 */

/**
 * Validate campaign form. Accepts the form values object and returns errors object.
 *
 * @param {CampaignFormValues} values Campaign form values.
 * @param {Object} opts Extra form options.
 * @return {Object} errors.
 */
const validateCampaign = ( values, opts ) => {
	const errors = {};

	if ( ! Number.isFinite( values.amount ) || values.amount <= 0 ) {
		errors.amount = __(
			'Please make sure daily average cost is greater than 0.',
			'google-listings-and-ads'
		);
	}

	if (
		Number.isFinite( values?.amount ) &&
		Number.isFinite( opts?.budget?.daily_budget )
	) {
		const { amount } = values;
		const { budget, budgetMin } = opts;
		const { daily_budget: dailyBudget } = budget;
		const minAmount = Math.round( dailyBudget * budgetMin, 2 );

		if ( amount < minAmount ) {
			errors.amount = sprintf(
				/* translators: %s: minimum daily budget */
				__(
					'Please make sure daily average cost is at least %s.',
					'google-listings-and-ads'
				),
				minAmount
			);
		}
	}

	return errors;
};

export default validateCampaign;
