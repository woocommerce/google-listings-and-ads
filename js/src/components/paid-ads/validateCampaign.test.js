/**
 * Internal dependencies
 */
import validateCampaign from './validateCampaign';

/**
 * `validateCampaign` function returns an object, and if any checks are not passed,
 * set properties respectively with an error message to indicate it.
 */
describe( 'validateCampaign', () => {
	let values;
	const validateCampaignOptions = {
		dailyBudget: undefined,
		formatAmount: jest.mock(),
	};

	beforeEach( () => {
		// Initial values
		values = { amount: 0 };
	} );

	it( 'When all checks are passed, should return an empty object', () => {
		const errors = validateCampaign(
			{
				amount: 1,
			},
			validateCampaignOptions
		);

		expect( errors ).toStrictEqual( {} );
	} );

	it( 'should indicate multiple unpassed checks by setting properties in the returned object', () => {
		const errors = validateCampaign( values, validateCampaignOptions );

		expect( errors ).toHaveProperty( 'amount' );
	} );

	it( 'When the amount is not a number, should not pass', () => {
		let errors;

		values.amount = '';
		errors = validateCampaign( values, validateCampaignOptions );

		expect( errors ).toHaveProperty( 'amount' );
		expect( errors.amount ).toMatchSnapshot();

		values.amount = undefined;
		errors = validateCampaign( values, validateCampaignOptions );

		expect( errors ).toHaveProperty( 'amount' );
		expect( errors.amount ).toMatchSnapshot();

		values.amount = new Date();
		errors = validateCampaign( values, validateCampaignOptions );

		expect( errors ).toHaveProperty( 'amount' );
		expect( errors.amount ).toMatchSnapshot();

		values.amount = NaN;
		errors = validateCampaign( values, validateCampaignOptions );

		expect( errors ).toHaveProperty( 'amount' );
		expect( errors.amount ).toMatchSnapshot();
	} );

	it( 'When the amount is â‰¤ 0, should not pass', () => {
		let errors;

		values.amount = 0;
		errors = validateCampaign( values, validateCampaignOptions );

		expect( errors ).toHaveProperty( 'amount' );
		expect( errors.amount ).toMatchSnapshot();

		values.amount = -0.01;
		errors = validateCampaign( values, validateCampaignOptions );

		expect( errors ).toHaveProperty( 'amount' );
		expect( errors.amount ).toMatchSnapshot();
	} );

	it( 'When a budget is provided and the amount is less than the minimum, should not pass', () => {
		const mockFormatAmount = jest.fn().mockReturnValue( 'Rs 30' );
		values.amount = 10;

		const opts = {
			dailyBudget: 100,
			formatAmount: mockFormatAmount,
		};

		const errors = validateCampaign( values, opts );

		expect( errors ).toHaveProperty( 'amount' );
		expect( errors.amount ).toContain( 'is at least Rs 30' );
	} );

	it( 'When a budget is provided and the amount is same than the minimum, should pass', () => {
		const mockFormatAmount = jest.fn().mockReturnValue( 'Rs 30' );
		values.amount = 30;

		const opts = {
			dailyBudget: 100,
			formatAmount: mockFormatAmount,
		};

		const errors = validateCampaign( values, opts );
		expect( errors ).not.toHaveProperty( 'amount' );
	} );

	it( 'When a budget is provided and the amount is greater than the minimum, should pass', () => {
		const mockFormatAmount = jest.fn().mockReturnValue( 'Rs 35' );
		values.amount = 35;

		const opts = {
			dailyBudget: 100,
			formatAmount: mockFormatAmount,
		};

		const errors = validateCampaign( values, opts );
		expect( errors ).not.toHaveProperty( 'amount' );
	} );
} );
