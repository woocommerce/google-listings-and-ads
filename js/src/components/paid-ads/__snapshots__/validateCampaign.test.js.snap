// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateCampaign When the amount is not a number, should not pass 1`] = `"Please make sure daily average cost is greater than 0."`;

exports[`validateCampaign When the amount is not a number, should not pass 2`] = `"Please make sure daily average cost is greater than 0."`;

exports[`validateCampaign When the amount is not a number, should not pass 3`] = `"Please make sure daily average cost is greater than 0."`;

exports[`validateCampaign When the amount is not a number, should not pass 4`] = `"Please make sure daily average cost is greater than 0."`;

exports[`validateCampaign When the amount is ≤ 0, should not pass 1`] = `"Please make sure daily average cost is greater than 0."`;

exports[`validateCampaign When the amount is ≤ 0, should not pass 2`] = `"Please make sure daily average cost is greater than 0."`;
